### DISCUSSION TITLE
Free library idea: drawing maps and geodataframes in py5

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#344] [Ideas]
I know it can sound a bit ridiculous, but, maybe if someone with the right skills comes by...

I noticed that geocoded data in a geodataframe can contain shapely geometry, and previously I have played with [Unfolding Maps in Processing.py](https://github.com/villares/material-aulas/tree/d7f721347ca30f465cbc8e8455857aba6c1809d4/exemplos-de-bibliotecas/UnfoldingMaps) so, some geo/map viz would be great to have.

**Comment by hx2A** [comment]
Yes, definitely, I'm interested in exploring map data and py5. I've wanted to do this for a long time but simply haven't had the time. Making progress with py5+shapely would be a step in the right direction, and make it easier for someone else to take the next step do something with mapping data.

**Comment by villares** [comment]
If I had the time I would study @marceloprates's [**prettymaps**](https://github.com/marceloprates/prettymaps/) code.

![image](https://github.com/py5coding/py5generator/assets/3694604/e4cf9ab5-f1d2-411c-84f0-2eaa4ae7ebe1)

**Reply by marceloprates** [reply]
All the geometries drawn with prettymaps are shapely objects! The main difference is that prettymaps uses matplotlib for plotting, but it shouldn't be too difficult to add a py5 drawing option. prettymaps currently supports two modes: matplotlib and vsketch (for pen plotters).

**Reply by hx2A** [reply]
Hello, @marceloprates ! If I added an easy way for py5 to work with shapely objects, would you be interested in collaborating to add a py5 mode for prettymaps?

One big benefit of this that I see is it would allow prettymaps users to use shaders. It also would enable people to incorporate it into animations.

**Reply by hx2A** [reply]
Last night I started working on this and have something basically working for Polygons and MultiPolygons. However, my goal of creating a convert method to convert Shapely objects to Py5Shape objects hit a snag: Py5Shape's `begin_contour()` method doesn't work for the default renderer. There is an [open Processing bug](https://github.com/processing/processing4/issues/643) providing more detail. I'll take look at the code to see if I can fix it, but it seems unlikely I'll fix it if nobody else has been able to. More likely is I adjust plans and add a `draw_shapely()` method to Sketch and Py5Graphics.

In addition, I want to point out that Shapely and py5 use different coordinate systems. Py5 has the origin in the upper left corner and positive Y is down. Shapely has origin in the lower left and positive Y is up. @villares , we'll need to make some good design choices to make sure users don't have the cognitive load of 2 coordinate systems.

**Reply by villares** [reply]
Hmm, do you have a link for the bug? I have drawn some holes with begin_contour()... There is the counter-clockwise requirement, etc... oh, I see, the .begin_contour() method on PShape/Py5Shape objects.

It never occurred to me shapely had an intrinsic coordinate system orientation...

**Reply by villares** [reply]
![image](https://github.com/py5coding/py5generator/assets/3694604/5381b0c4-4e31-4555-a797-8bf53b01d536)

![image](https://github.com/py5coding/py5generator/assets/3694604/0c030852-a9b6-4c8a-9eb1-a46d4ded057f)

**Reply by hx2A** [reply]
> It never occurred to me shapely had an intrinsic coordinate system orientation...

At first glance it doesn't, but if you convert a shape to SVG with the `svg()` method you will see it will draw things differently from how py5 does it. More importantly, consider transformations, and in particular, rotations. A clockwise rotation to a shapely shape should yield a clockwise rotation to what py5 draws.

**Reply by hx2A** [reply]
> Hmm, do you have a link for the bug? I have drawn some holes with begin_contour()... There is the counter-clockwise requirement, etc... oh, I see, the .begin_contour() method on PShape/Py5Shape objects.

I figured out the problem and have a fix for it. It's not a big change. After doing some testing, I'll make a PR.

**Reply by hx2A** [reply]
Also, here's a link to the open issue:

https://github.com/processing/processing4/issues/643

I pasted in the wrong link before.

I'll make a PR if my fix holds up.

**Reply by hx2A** [reply]
I made the PR - fingers crossed it gets merged

**Reply by marceloprates** [reply]
Hey @hx2A ! Sure, I'd happy to! I don't have much free time at the moment, but could collaborate on that as soon as I finish some other activities.

**Reply by hx2A** [reply]
@marceloprates , I'm looking forward to working on this! I'm currently adding a feature to py5 that we'll want to use to interface your code to py5. I'll let you know when that is ready.

**Reply by marceloprates** [reply]
Awesome, please let me know when it's ready!

**Comment by villares** [comment]
My crude explorations, some progress  :)

![sketch_2023_08_27](https://github.com/py5coding/py5generator/assets/3694604/4288d0b3-2ff1-4230-9c9d-e095aabe7772)

```python
# Draws ways and buildings from data © OpenStreetMap contributors
import py5

from shapely import Polygon, MultiPolygon
from shapely import LineString, MultiLineString, LinearRing
from shapely import Point, MultiPoint
from shapely import GeometryCollection
from shapely.ops import transform

import osmnx as ox
ox.settings.log_console=True
ox.settings.use_cache=True

from geopandas import GeoDataFrame

bgcolor = '#343434'
edge_color = '#CCCCCC'
bldg_color = '#008888'
point = ox.geocode('Avenida Paulista, São Paulo, Brasil')
dist = 2000

bbox = ox.utils_geo.bbox_from_point(point, dist=dist)
fp = ox.features_from_point(point, tags={'building':True}, dist=dist)
G = ox.graph_from_point(point, network_type='walk', dist=dist, truncate_by_edge=True, retain_all=True)
nodes, edges = ox.graph_to_gdfs(G)
m_edges = edges.to_crs(epsg=3857)
m_fp = fp.to_crs(epsg=3857)
#m_edges = edges.set_crs('EPSG:4326').to_crs('EPSG:3857')
paulista = m_edges[m_edges['name'] == 'Avenida Paulista']
#selected_edges_gdf = gdf_meters[gdf_meters['edge_id'].isin(selected_edges)]
paulista_length = paulista.geometry.length.sum()
print(paulista_length)  # value seems wrong!!!

# for debugging:
import matplotlib.pyplot as plt
fig, ax = ox.plot_graph(G, bgcolor=bgcolor, node_size=0, edge_color=edge_color, show=False)
fig, ax = ox.plot_footprints(fp, ax=ax, bbox=bbox, color=bldg_color, show=False) #, save=True)
paulista.plot(ax=ax, color='red')
# an then call plt.show() at the REPL...

def setup():
    py5.size(800, 800)
    py5.background(bgcolor)
    #py5.translate(400, 400)
    minx, miny, maxx, maxy = m_edges.total_bounds
    scale_x = py5.width / (maxx - minx)
    scale_y = py5.height / (maxy - miny)
    translate_x = -minx * scale_x
    translate_y = -miny * scale_y
    matrix = [scale_x, 0, 0, scale_y, translate_x, translate_y]
    screen_edges = m_edges['geometry'].affine_transform(matrix).apply(flip_y)
    screen_buildings = m_fp['geometry'].affine_transform(matrix).apply(flip_y)
    screen_paulista = paulista.affine_transform(matrix).apply(flip_y)
    py5.no_stroke()
    py5.fill(bldg_color)
    draw_shapely(b for b in screen_buildings if b.area)    
    py5.stroke(edge_color)
    draw_shapely(screen_edges)
    py5.stroke(255, 0, 0)
    draw_shapely(screen_paulista)
    py5.save(f'{__file__[:-3]}.png')

def flip_y(geom):
    return transform(lambda x, y: (x, -y + py5.height), geom)
 
def draw_shapely(shps, sketch: py5.Sketch=None):
    """
    Draw most shapely objects with py5.
    This will use the "current" py5 sketch as default.
    """
    s = sketch or py5.get_current_sketch()
    if isinstance(shps, (MultiPolygon, MultiLineString, GeometryCollection)):
        for shp in shps.geoms:
            draw_shapely(shp)
    elif isinstance(shps, Polygon):
        with s.begin_closed_shape():
            s.vertices(shps.exterior.coords)
            for hole in shps.interiors:
                with s.begin_contour():
                    s.vertices(hole.coords)
    elif isinstance(shps, (LineString, LinearRing)):
        # no need to uses begin_closed_shape() because LinearRing repeats the start/end coordinates
        with s.push_style():
            s.no_fill()
            with s.begin_shape():
                s.vertices(shps.coords)
    elif isinstance(shps, Point):
        s.point(shps.x, shps.y) 
    elif isinstance(shps, MultiPoint):
        s.points((p.x, p.y) for p in shps.geoms)
    elif isinstance(shps, GeoDataFrame):
        for shp in shps.geometry:
                draw_shapely(shp)
    else:
        try:
            for shp in shps:
                draw_shapely(shp)
        except TypeError as e:
            print(f"Unable to draw: {shps}")
        
py5.run_sketch(block=False)
```

**Reply by villares** [reply]
Flipping stuff upfront seems costly, so the example linked below would bring it into the (rudimentary) drawing function, but,  as @hx2A anticipated, thinking in 2 coordinates systems at once is hard... 

https://github.com/villares/sketch-a-day/blob/main/2023/sketch_2023_08_28/sketch_2023_08_28.py

**Reply by hx2A** [reply]
> Flipping stuff upfront seems costly, so the example linked below would bring it into the (rudimentary) drawing function, but, as @hx2A anticipated, thinking in 2 coordinates systems at once is hard...

My current strategy is to alter the Shapely shape to convert it's coordinate system. Something like

```python
from shapely import affinity

obj = affinity.scale(obj, yfact=-1, origin="center")
```

I'm not sure if that will most correctly address the issue though, need more time with this.

In any case, I have a complete prototype of the convert functions for all Shapely geometry types. I need Processing to accept my PR though.

**Reply by hx2A** [reply]
> In any case, I have a complete prototype of the convert functions for all Shapely geometry types. I need Processing to accept my PR though.

Things are moving along here. Processing accepted my PR and I have merged the py5 branch with the shapely and trimesh code. This week I'll take a look at the prettymaps code to get ideas of how py5 and prettymaps can work together and I'll report back.

**Comment by Maltiec** [comment]
Hmm, I have some GPS data of my bicycle riding in a city park - may be it's time to build a map of bicycle routes.)

**Reply by hx2A** [reply]
That'd be a great idea!

**Comment by hx2A** [comment]
Here's a map made in py5 with prettymaps:

![image](https://github.com/py5coding/py5generator/assets/4044283/067d3896-5137-484f-aae9-0c49e6b71eb0)

Do you recognize the location?

This is a prototype but it is in pretty good shape. While building this I found a bug in Processing and made a PR and also found a bug in py5's new shapely integration code. Working on this has been a useful test of the new py5 feature.

@marceloprates , the plotter / vsketch mode seems to be commented out and is not working? I need to support the "background" functionality and maybe some other things I am missing. The py5 code is still a bit messy and I will make sure there are no bugs on my end before asking you to do anything.

py5 does not have hatching so we'll want to make some presets that are well suited for py5. I'll enlist @villares and others for help with that when the prototype is ready for it.

**Reply by villares** [reply]
wow... [Francesca Franco](https://www.francescafranco.net/) would have enjoyed this I guess ;-)

**Reply by hx2A** [reply]
you got it!
